src/core/qgsvectorlayerfeatureiterator.cpp

```diff
@@ -140,20 +140,38 @@ QgsVectorLayerFeatureIterator::QgsVectorLayerFeatureIterator( QgsVectorLayerFeat
     mRequest.setFilterRect( mFilterRect );
   }
 
+  QSet<int> attributeIndexes;
+
+  // check whether the order by clause(s) can be delegated to the provider
+  mDelegatedOrderByToProvider = !mSource->mHasEditBuffer;
+  if ( !mRequest.orderBy().isEmpty() )
+  {
+    const auto usedAttributeIndices = mRequest.orderBy().usedAttributeIndices( mSource->mFields );
+    for ( int attrIndex : usedAttributeIndices )
+    {
+      if ( mSource->mFields.fieldOrigin( attrIndex ) != QgsFields::OriginProvider )
+        mDelegatedOrderByToProvider = false;
+
+      attributeIndexes << attrIndex;
```

Comment:
We also need to check the various added/edited/deleted feature maps are empty. If not, we can't delegate to provider...

Reply:
This is happening on line 146, i.e. `mDelegatedOrderByToProvider = !mSource->mHasEditBuffer;`